### PR TITLE
Solution submission step4: Introducing fee mechanism 

### PR DESCRIPTION
**Pull request by josojo**
This Pr introduces the fee mechanism as discussed here: https://github.com/gnosis/dex-contracts/issues/173.

Now, the constructor takes a fee parameter and a fee token. The fee token will be the token with the token index 0 and must be part of any solution. 

In the solution vector, the tokenIdsForIndex array does not need to supply the feeToken index 0, as it must be part of any solution and hence we implicitly know that the first index would always have to be pointing to 0 anyways. 

Every solution must be constructed such that the imbalance of tokens bought and tokens sold is shifted into the fee token. Hence, the token conservation check is only done for all tokens, but the fee token.

-----
I was rewriting most of the tests to base them on a basic trade, as the fee mechanism required me to do this anyways. Now, we no longer have to check all the numbers in the tests, we just reuse the same numbers for the basic trade and modify them to archive specific test results.

---
Currently, this Pr is super long. If you do not want to review it, I can understand. Please let me know how I could do the test refactoring more efficient and more comprehensive.

---
Test plan: Unit tests only

### CODE REVIEW

File: contracts/StablecoinConverter.sol [hunk 1]
```diff
@@ -175,18 +186,35 @@ contract StablecoinConverter is EpochTokenLocker {
                 volumes[i]
             );
         }
+        int fee = tokenConservation[0];
+        require(fee < auctionData[batchIndex].feeCollected, "Fee is not higher than before");
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Should this be strictly higher as long as the current one is > 0? Otherwise I can replace existing solutions without improving them.

File: contracts/StablecoinConverter.sol [hunk 2]
```diff
@@ -112,8 +115,14 @@ contract StablecoinConverter is EpochTokenLocker {
     }
 
     mapping (uint16 => uint128) public currentPrices;
+    mapping (uint32 => AuctionData) public auctionData;
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Why do we store this with history?

File: contracts/StablecoinConverter.sol [hunk 3]
```diff
@@ -175,18 +186,35 @@ contract StablecoinConverter is EpochTokenLocker {
...
+        auctionData[batchIndex].bestSolutionSubmitter = msg.sender;
         checkTokenConservation(tokenConservation);
         documentTrades(batchIndex, owners, orderIds, volumes, tokenIdsForPrice);
     }
 
     function checkTokenConservation(
         int[] memory tokenConservation
     ) internal pure {
-        for (uint i = 0; i < tokenConservation.length; i++) {
+        for (uint i = 1; i < tokenConservation.length; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol:218]: You probably still want to check that for fee token we have tokenConservation > 0?

File: contracts/StablecoinConverter.sol [hunk 4]
```diff
@@ -155,16 +165,17 @@ contract StablecoinConverter is EpochTokenLocker {
                 currentPrices[order.buyToken],
                 currentPrices[order.sellToken]
             );
+            executedBuyAmount = uint128(executedBuyAmount.mul(feeDenominator - 1) / feeDenominator);
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: It took me a while fully understand this logic here. Maybe it should be documented in the code.

If `fee = 1/denominator` and according to [Tom's PR](https://gitlab.gnosisdev.com/dfusion/batchauctions/merge_requests/1/diffs#f64fa3d7c092c29083770d4926252ca0ffe4eae1_899_901) `executedBuyAmount = buyAmount * (1 - fee)`

Then

```
executedBuyAmount = buyAmount * (1 - (1/feeDenominator)
                  = buyAmount - buyAmount/feeDenominator (*)
                  = (buyAmount * feeDenominator)/ feeDenominator - buyAmount/feeDenominator
                  = (buyAmount * feeDenominator - buyAmount) / feeDenominator
                  = (buyAmount* (feeDenominator - 1)/feeDenominator
```

(*) Is there any reason we are not using this formula?


Also, do we want to handle the case where fee is 0? This could also simplify a lot of your unit tests.

File: contracts/StablecoinConverter.sol [hunk 5]
```diff
@@ -175,18 +186,35 @@ contract StablecoinConverter is EpochTokenLocker {
...
+        for (uint i = 1; i < tokenConservation.length; i++) {
             require(tokenConservation[i] == 0, "Token conservation does not hold");
         }
     }
 
+    function updateCurrentPrices(
+        uint128[] memory prices,  //list of prices for touched token only, frist price is fee Token price
+        uint16[] memory tokenIdsForPrice  // price[i] is the price for the token with tokenID tokenIdsForPrice[i]
+    ) internal {
+        for (uint i = 0; i < previousSolution.tokenIdsForPrice.length; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: This looks like it contains a bugfix. Should this be a separate PR with an added unit test?

File: contracts/StablecoinConverter.sol [hunk 6]
```diff
@@ -175,18 +186,35 @@ contract StablecoinConverter is EpochTokenLocker {
...
+            currentPrices[previousSolution.tokenIdsForPrice[i]] = 0;
+        }
+        currentPrices[0] = prices[0];
+        for (uint i = 0; i < tokenIdsForPrice.length; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol:227]: Does it make sense to start from 0 here?

File: contracts/StablecoinConverter.sol [hunk 7]
```diff
@@ -260,31 +279,39 @@ contract StablecoinConverter is EpochTokenLocker {
                     currentPrices[order.buyToken],
                     currentPrices[order.sellToken]
                 );
+                buyVolume = uint128(buyVolume.mul(feeDenominator - 1) / feeDenominator);
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Should this logic be part of `getExecutedBuyAmount`?

File: contracts/StablecoinConverter.sol [hunk 8]
```diff
@@ -260,31 +279,39 @@ contract StablecoinConverter is EpochTokenLocker {
...
             }
         }
         revert("Price not provided for token");
     }
 
-    function checkPriceOrdering(uint16[] memory tokenIdForPrice) private pure returns (bool) {
-        for (uint i = 1; i < tokenIdForPrice.length; i++) {
-            if (tokenIdForPrice[i] <= tokenIdForPrice[i - 1]) {
+    function checkPriceOrdering(uint16[] memory tokenIdsForPrice) private pure returns (bool) {
+        require(tokenIdsForPrice[0] > 0, "price for fee token should not be overwritten");
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: I don't understand this require. Shouldn't it say something like "fee token price index does not have to be specified"?

File: contracts/StablecoinConverter.sol [hunk 9]
```diff
@@ -260,31 +279,39 @@ contract StablecoinConverter is EpochTokenLocker {
...
         uint rightValue = tokenIdForPrice.length - 1;
         while (rightValue >= leftValue) {
             uint middleValue = leftValue + (rightValue-leftValue) / 2;
             if (tokenIdForPrice[middleValue] == index) {
-                return middleValue;
+                return middleValue + 1;
             } else if (tokenIdForPrice[middleValue] < index) {
                 leftValue = middleValue + 1;
             } else {
+                require(middleValue > 0, "Price not provided for token, underflow would have happened");
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Is this really required? Wouldn't we just do one more iteration and then fail on the existing require?

File: contracts/StablecoinConverter.sol [hunk 10]
```diff
@@ -133,6 +142,7 @@ contract StablecoinConverter is EpochTokenLocker {
         uint128[] memory volumes,
         uint128[] memory prices,  //list of prices for touched token only
         uint16[] memory tokenIdsForPrice  // price[i] is the price for the token with tokenID tokenIdsForPrice[i]
+                                          // tokenId for fee token does not need to be send, as it is always 0
```
Comment (bh2smith) [inline contracts/StablecoinConverter.sol]: I am not sure I understand this comment. May I suggest the following if my interpretation is correct?
```suggestion
                                          // fee token id not required since always 0
```

Comment (fleupold) [inline contracts/StablecoinConverter.sol]: I'm not sure if the added complexity here (shifting binary search by one, requiring 0 to not be part of the tokenId list) is worth the one uint16 calldata saving. I'd instead just assert that the first entry is 0 and keep the remaining logic as is.

**Review by fleupold — COMMENTED** [review]
Looks good, but some questions and simplification suggestions inline. Couldn't review the unit tests since they are too big.

File: contracts/StablecoinConverter.sol (hunk 1 above)
Comment (bh2smith) [inline contracts/StablecoinConverter.sol]: I am not sure this error message is complete. "Before" what?

Maybe something like "Collected fee must exceed fee requested"?

File: contracts/StablecoinConverter.sol [hunk 11]
```diff
@@ -175,18 +186,35 @@ contract StablecoinConverter is EpochTokenLocker {
...
         int[] memory tokenConservation
     ) internal pure {
-        for (uint i = 0; i < tokenConservation.length; i++) {
+        for (uint i = 1; i < tokenConservation.length; i++) {
             require(tokenConservation[i] == 0, "Token conservation does not hold");
         }
     }
 
+    function updateCurrentPrices(
+        uint128[] memory prices,  //list of prices for touched token only, frist price is fee Token price
```
Comment (bh2smith) [inline contracts/StablecoinConverter.sol]: nit: could we align this comment with the one in the next line? Also, "touched tokens" - plural.

File: contracts/StablecoinConverter.sol (hunk 1 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: The way we are calculating the tokenConservation, it will always be smaller than 0. That means the fee is a negative number. We are selling more than we are buying. Is this confusing? Probably yes, let me think of a solution.

Yes, it is a strict inequality

File: contracts/StablecoinConverter.sol (hunk 2 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: auctionData[batchIndex].feeCollected is needed to judge new solutions. I added the msg.sender, as it is important later for the withdraw. Yeah, I can take the bestSolutionSubmitter out, as it is only needed later for the withdraw of the reward

File: contracts/StablecoinConverter.sol (hunk 3 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol:218]: No, this is not needed. If the token conservation holds for all other token pairs, and we have arbitrage free prices + fees, then it must hold that tokenConservation < 0 for the fee token.

File: contracts/StablecoinConverter.sol (hunk 2 above)
Comment (fleupold) [reply contracts/StablecoinConverter.sol]: Makes sense that we need history for later reward withdrawal (we probably don't need history on the fees though).
Would it make sense to pay the reward as soon as the first solution in the next batch is submitted?

Comment (fleupold) [reply contracts/StablecoinConverter.sol]: Actually speaking about rewards, isn't the reward that the solver credits themselves the fee surplus? This would automatically get reverted by a better solution so there shouldn't be a need to later claim rewards.

That being said, the part about crediting the difference is not yet implemented, but in my opinion it should be part of "apply solution".

Comment (josojo) [reply contracts/StablecoinConverter.sol]: And we store it with history, as it is needed for the GNO withdrawals later from the solution submitter. Once the withdraw is done, we can set the history value to 0

File: contracts/StablecoinConverter.sol [hunk 12]
```diff
@@ -260,31 +279,39 @@ contract StablecoinConverter is EpochTokenLocker {
...
             }
         }
     }
 
     function findPriceIndex(uint16 index, uint16[] memory tokenIdForPrice) private pure returns (uint) {
+        // return fee token
+        if (index == 0) {
+            return 0;
+        }
+        // binary search for the other tokens
```
Comment (bh2smith) [inline contracts/StablecoinConverter.sol:317]: Very nice! When did this happen! reduction from O(n) to O(log n)! Sweet

File: contracts/StablecoinConverter.sol (hunk 9 above)
Comment (bh2smith) [inline contracts/StablecoinConverter.sol]: In what scenario does this happen?

File: contracts/StablecoinConverter.sol (hunk 4 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: Cool, I will add this comment to the code base.

>(*) Is there any reason we are not using this formula?

yeah, we don't have decimal numbers, so we represent 0.999 as *999/1000

>Also, do we want to handle the case where fee is 0? 

Yes, it would make the unit tests much simpler, but then the code in the smart contract would be more complex. I decided for complexity in the tests over complexity in the smart contract itself

File: contracts/StablecoinConverter.sol (hunk 8 above)
Comment (bh2smith) [inline contracts/StablecoinConverter.sol]: Why does being zero imply it was overwritten?

File: contracts/StablecoinConverter.sol (hunk 5 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: Yeah, let me separate it out.

File: contracts/StablecoinConverter.sol (hunk 6 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol:227]: Yes, as we also want to cover: tokenIdsForPrice[0], which is not the fee token

File: contracts/StablecoinConverter.sol (hunk 8 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: Yeah, your wording is better.
But the effect of the specification `tokenIdsForPrice[0]` would be that the fee token price is indeed overwritten:

```
function updateCurrentPrices(
        uint128[] memory prices,  //list of prices for touched token only, frist price is fee Token price
        uint16[] memory tokenIdsForPrice  // price[i] is the price for the token with tokenID tokenIdsForPrice[i]
    ) internal {
        for (uint i = 0; i < previousSolution.tokenIdsForPrice.length; i++) {
            currentPrices[previousSolution.tokenIdsForPrice[i]] = 0;
        }
        currentPrices[0] = prices[0];
        for (uint i = 0; i < tokenIdsForPrice.length; i++) {
            currentPrices[tokenIdsForPrice[i]] = prices[i + 1];
        }
    }
```

File: contracts/StablecoinConverter.sol (hunk 9 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: We would calculate middleValue -1 = 0-1 and it would underflow. I prefer to have an explicit error message than a normal underflow error.

File: test/stablecoin_converter.js [hunk 13]
```diff
@@ -16,14 +21,30 @@ contract("StablecoinConverter", async (accounts) => {
...
 
-    const stablecoinConverter = await StablecoinConverter.new(2 ** 16 - 1)
     BATCH_TIME = (await stablecoinConverter.BATCH_TIME.call()).toNumber()
   })
 
+  // Basic Trade used in most of the tests:
+  // Trade for user_1: amount of token_1 sold: 20000, amount of token_2 bought: feeSubtracted(10000),
+  // Trade for user_2: amount of token_2 sold: feeSubtracted(10000), amount of token_1 bought: feeSubtracted(feeSubtracted(10000) * 2)
+  // ==> Token conservation holds for token_2, and fee token == token_1 has negative balance of 40
+  const basicTrade = {
```
Comment (bh2smith) [inline test/stablecoin_converter.js:36]: This is cool!

File: contracts/StablecoinConverter.sol (hunk 10 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: Yeah, I also did not implement that immediately. But since we are also storing the tokenIdsForPrice, I thought that is worth the complexity. But for sure, you have a point here.

**Review by bh2smith — APPROVED** [review]
This looks pretty great! Several inline comments and questions probably mostly concerned with a few `require` statements that I was unsure of.

File: contracts/StablecoinConverter.sol (hunk 2 above)
Comment (fleupold) [reply contracts/StablecoinConverter.sol]: Why don't we credit the GNO to the msg.sender's balance just like any other trade execution?

File: contracts/StablecoinConverter.sol (hunk 4 above)
Comment (fleupold) [reply contracts/StablecoinConverter.sol]: If this logic becomes part of getExecutedBuyAmount I think it's worth adding this complexity since it's only in a single case.

File: contracts/StablecoinConverter.sol (hunk 10 above)
Comment (fleupold) [reply contracts/StablecoinConverter.sol]: Overall there are too many (5) confusing small changes that are all related to saving a tiny amount of data here (that will also prevent us from using standard library implementations).

File: contracts/StablecoinConverter.sol (hunk 4 above)
Comment (bh2smith) [reply contracts/StablecoinConverter.sol]: I recall a time in the past where we initially opted for less contract complexity (namely how the zeroth slot is treated) and eventually landed on moving the complexity into the contract to make the end-to-end tests more coherent. Is this something we expect to change here in the future as well?

File: contracts/StablecoinConverter.sol (hunk 5 above)
Comment (bh2smith) [reply contracts/StablecoinConverter.sol]: Is this the `+1` introduced in the binary search? It seems like the comment is on the wrong line. If so, I would agree this looks like a different issue.

File: contracts/StablecoinConverter.sol (hunk 9 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: Yeah, it's not a usual case. Check out the unit test for it.

File: contracts/StablecoinConverter.sol (hunk 12 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol:317]: Felix was pushing for it :)

File: contracts/StablecoinConverter.sol (hunk 8 above)
Comment (josojo) [reply contracts/StablecoinConverter.sol]: In the price vector, the first element is reserved for the fee token. If another index is also pointing to 0, then there are two prices supplied for the fee token. The second one would might overwrite the first input then

Comment (josojo) [reply contracts/StablecoinConverter.sol]: But as Felix suggested, I changed the message

**Comment by fleupold** [comment]
What's your thought on all these one-off changes in order to save one price (uint16) of calldata? 
I think undoing those would greatly help readability, maintainability and potential to replace logic with standard library implementations going forward.

If it's not done in this PR, it will also require changes in the driver if it firsts adapt to this API.

File: contracts/StablecoinConverter.sol [hunk 14]
```diff
@@ -219,11 +216,17 @@ contract StablecoinConverter is EpochTokenLocker {
...
+    ) internal view returns (uint128) {
+        uint128 buyAmount = uint128(
             executedSellAmount.mul(buyTokenPrice) /
             sellTokenPrice
         );
+        // executedBuyAmount = buyAmount * (1 - (1/feeDenominator)
+        //           = buyAmount - buyAmount/feeDenominator (*)
+        //           = (buyAmount * feeDenominator)/ feeDenominator - buyAmount/feeDenominator
+        //           = (buyAmount * feeDenominator - buyAmount) / feeDenominator
+        //           = (buyAmount* (feeDenominator - 1)/feeDenominator
```
Comment (bh2smith) [inline contracts/StablecoinConverter.sol]: Should these commented lines be removed?

Comment (josojo) [reply contracts/StablecoinConverter.sol]: Felix asked for them :) 
I think anything that helps readability is a plus

**Comment by josojo** [comment]
>What's your thought on all these one-off changes in order to save one price (uint16) of calldata?
I think undoing those would greatly help readability, maintainability and potential to replace logic with standard library implementations going forward.

Okay, I will do it, as I am undecided and don't have a strong opinion.
But we should be aware that it is not just calldata, as the first tokenIdsForPrices have to be stored,

File: contracts/StablecoinConverter.sol [hunk 15]
```diff
@@ -175,27 +185,43 @@ contract StablecoinConverter is EpochTokenLocker {
...
         uint128 sellTokenPrice
-    ) internal pure returns (uint128) {
-        return uint128(
+    ) internal view returns (uint128) {
+        uint128 buyAmount = uint128(
             executedSellAmount.mul(buyTokenPrice) /
             sellTokenPrice
         );
+        // executedBuyAmount = buyAmount * (1 - (1/feeDenominator)
+        //           = buyAmount - buyAmount/feeDenominator (*)
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: uber nit: would be nice if the `=` align all the way and maybe become `<=>`

File: contracts/StablecoinConverter.sol [hunk 16]
```diff
@@ -241,32 +258,36 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                sellVolume,
+                currentPrices[order.buyToken],
+                currentPrices[order.sellToken]
+            );
+            revertRemainingOrder(owner, orderId, sellVolume);
+            subtractBalance(owner, tokenIdToAddressMap(order.buyToken), buyVolume);
+        }
+    }
+
+    function checkForBestSolutionSubmission(int256 fee) private {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Not a big fan of this helper method. It does both a check and a side-effect which is not very obvious from the name. Maybe checkAndOverrideObjectiveValue?

File: contracts/StablecoinConverter.sol [hunk 17]
```diff
@@ -241,32 +258,36 @@ contract StablecoinConverter is EpochTokenLocker {
...
-                uint128 buyVolume = getExecutedBuyAmount(
-                    sellVolume,
-                    currentPrices[order.buyToken],
-                    currentPrices[order.sellToken]
-                );
-                revertRemainingOrder(owner, orderId, sellVolume);
-                subtractBalance(owner, tokenIdToAddressMap(order.buyToken), buyVolume);
-            }
+    function undoPreviousSolution() internal {
+        for (uint i = 0; i < previousSolution.trades.length; i++) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Is this change really related to introducing a fee mechanism?

File: contracts/StablecoinConverter.sol [hunk 18]
```diff
@@ -175,27 +185,43 @@ contract StablecoinConverter is EpochTokenLocker {
...
     ) internal pure {
-        for (uint i = 0; i < tokenConservation.length; i++) {
+        for (uint i = 1; i < tokenConservation.length; i++) {
             require(tokenConservation[i] == 0, "Token conservation does not hold");
         }
     }
 
+    function updateCurrentPrices(
+        uint128[] memory prices,          // list of prices for touched tokens only, frist price is fee token price
+        uint16[] memory tokenIdsForPrice  // price[i + 1] is the price for the token with tokenID tokenIdsForPrice[i]
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: This comment seems incorrect now

File: contracts/StablecoinConverter.sol [hunk 19]
```diff
@@ -133,13 +137,19 @@ contract StablecoinConverter is EpochTokenLocker {
         uint128[] memory volumes,
         uint128[] memory prices,  //list of prices for touched token only
         uint16[] memory tokenIdsForPrice  // price[i] is the price for the token with tokenID tokenIdsForPrice[i]
+                                          // fee token id not required since always 0
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: this comment seems incorrect now

File: contracts/StablecoinConverter.sol [hunk 20]
```diff
@@ -133,13 +137,19 @@ contract StablecoinConverter is EpochTokenLocker {
...
     ) public {
         require(
             batchIndex == getCurrentStateIndex() - 1,
             "Solutions are no longer accepted for this batch"
         );
         require(checkPriceOrdering(tokenIdsForPrice), "prices are not ordered by tokenId");
-        undoPreviousSolution(batchIndex);
+        if (previousSolution.batchId == batchIndex) {
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: I'd prefer if this branch stayed inside the sub-function.

File: contracts/StablecoinConverter.sol [hunk 21]
```diff
@@ -133,13 +137,19 @@ contract StablecoinConverter is EpochTokenLocker {
...
+            undoPreviousSolution();
+        } else {
+            currentFeeCollected = 0;
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: Shouldn't this logic be part of `checkForBestSolutionSubmission` given that it will be overridden in the same method anyways? Seems a bit dislocated here.

**Review by fleupold — APPROVED** [review]
Only some small clumsy things remaining.

**Review by anxolin — COMMENTED** [review]
Arrived late to the party, but looks good!!

File: contracts/StablecoinConverter.sol (hunk 12 above)
Comment (bh2smith) [reply contracts/StablecoinConverter.sol:317]: https://github.com/gnosis/dex-contracts/pull/164#pullrequestreview-273126587
